### PR TITLE
Fix issue where async as an id cannot be used with application and la…

### PR DESCRIPTION
**Pull request by cristianoc**
…belled arguments.

Fixes https://github.com/rescript-lang/syntax/issues/707

### CODE REVIEW

**Comment by cristianoc** [comment]
@IwanKaramazow would you check that there's nothing wrong with this fix in the heuristic for recognising es6 arrow functions? Is the early recognition of `~` just a perf shortcut, or does it have a deeper meaning?

File: tests/printer/expr/asyncAwait.res [hunk 1]
```diff
@@ -98,4 +98,7 @@ let f12 = @a (@b x) => 3
 let f13 = @a @b (~x) => 3
 
 let aw = (await (server->start))->foo
-let aw = (@foo (server->start))->foo
\ No newline at end of file
+let aw = (@foo (server->start))->foo
+
+let foo = async(~a=34)
```
Comment (IwanKaramazow) [inline tests/printer/expr/asyncAwait.res:103]: Might be more appropriate in the `tests/parsing/grammar/expressions/async` directory as it's really about parsing

**Review by IwanKaramazow — APPROVED** [review]

File: tests/printer/expr/asyncAwait.res (hunk 1 above)
Comment (cristianoc) [reply tests/printer/expr/asyncAwait.res:103]: Did that first but wanted to check that it does not mess up printing.
Probably both.

**Comment by IwanKaramazow** [comment]
> Is the early recognition of ~ just a perf shortcut, or does it have a deeper meaning?

Perf shortcut. There wasn't a case where `(~` didn't mean an arrow expression.

Your fix is correct. Upon finding the `async` keyword it will now search for the closing `)` and then look for an `=>` to confirm if we're in a arrow expression.

**Pull request merged**